### PR TITLE
Josue guessennd/cu 8685q56zj replace custom price attestation oracle with lookup to chainlink

### PR DESCRIPTION
**Pull request by joshualyguessennd**

### CODE REVIEW

**Comment by jaketimothy** [comment]
Task linked: [CU-8685q56zj Replace custom price attestation oracle with lookup to chainlink](https://app.clickup.com/t/8685q56zj)

File: src/forwarder/Forwarder.sol [hunk 1]
```diff
@@ -123,6 +120,16 @@ contract Forwarder is IForwarder, Ownable, PriceAttestationConsumer, Nonces, Mul
         emit CancellationGasCostUpdated(newCancellationGasCost);
     }
 
+    /**
+     * @dev add oracle for a payment token
+     * @param _paymentAsset asset to add oracle
+     * @param _oracle chainlink oracle address
+     */
+    function addOracle(address _paymentAsset, address _oracle) external onlyOwner {
```
Comment (jaketimothy) [inline src/forwarder/Forwarder.sol:128]: rename to updateOracle. we may want to remove oracles by setting value to 0

**Review by jaketimothy — CHANGES_REQUESTED** [review]

File: src/forwarder/IForwarder.sol [hunk 2]
```diff
@@ -1,22 +1,19 @@
...
 interface IForwarder {
     struct ForwardRequest {
         address user; // The address of the user initiating the meta-transaction.
         address to; // The address of the target contract (e.g., OrderProcessor)
             // to which the meta-transaction should be forwarded.
+        address paymentToken; // token use to pay transaction
         bytes data; // Encoded function call that the user wants to execute
             // through the meta-transaction.
-        uint64 deadline; // The time by which the meta-transaction must be mined.
+        uint256 deadline; // The time by which the meta-transaction must be mined.
```
[the remark this answers is not in the document]
Comment (joshualyguessennd) [reply src/forwarder/IForwarder.sol:14]: testing in the fork network gives an error when I set block.timestamp , I'm able to bypass with uint256 type.

File: src/forwarder/Forwarder.sol [hunk 3]
```diff
@@ -133,6 +140,20 @@ contract Forwarder is IForwarder, Ownable, PriceAttestationConsumer, Nonces, Mul
...
     }
 
+    /**
+     * @dev get the latest price of a token
+     * @param _asset asset to get the price
+     */
+    function getPaymentPriceInWei(address _asset) public view returns (uint256) {
+        address _oracle = paymentOracle[_asset];
+        // slither-disable-next-line unused-return
+        (, int256 paymentPrice,,,) = AggregatorV3Interface(_oracle).latestRoundData();
```
[the remark this answers is not in the document]
Comment (joshualyguessennd) [reply src/forwarder/Forwarder.sol:150]: Aggregator has already checks before retrieving the price, TX will failed if the price is wrong

File: src/forwarder/IForwarder.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (jaketimothy) [reply src/forwarder/IForwarder.sol:14]: pushing a fix for this

**Comment by jaketimothy** [comment]
@joshualyguessennd  Can you double check that the natspec and comments are still accurate with these new changes? The code looks good

File: src/forwarder/Forwarder.sol [hunk 4]
```diff
@@ -43,7 +43,7 @@ contract Forwarder is IForwarder, Ownable, Nonces, Multicall, SelfPermit, Reentr
     /// ------------------------------- Constants -------------------------------
 
     bytes private constant FORWARDREQUEST_TYPE = abi.encodePacked(
-        "ForwardRequest(address user,address to, address paymentToken, bytes data,uint256 deadline,uint256 nonce)"
+        "ForwardRequest(address user,address to,address paymentToken,bytes data,uint256 deadline,uint256 nonce)"
```
Comment (joshualyguessennd) [inline src/forwarder/Forwarder.sol:46]: the previous commit used uint64 for deadline

Comment (jaketimothy) [reply src/forwarder/Forwarder.sol]: good catch!

**Review by jaketimothy — APPROVED** [review]

**Pull request merged**